File: Loenn/entities/controllers/VectorHelperVariableController.lua
-- VectorHelper Variable Controller (Loenn Integration)
-- Written by SSUnlimited

local drawableSprite = require("structs.drawable_sprite")
local Selene = require("selene")

local vectorHelper = require("mods").requireFromPlugin("libraries.VectorHelper")
local consts = require("mods").requireFromPlugin("libraries.Consts")

local variableController = {}

variableController.name = "VectorHelper/VariableController"
variableController.depth = -15000
variableController.placements = {
	{
		name = "variable_controller",
		data = {
			type = "Variable",
			dataType = "String",
			variableName = "myVariable",
			initialValue = "",
			variableFor = "SaveData",
			oneTime = "False",
			arrayLength = "1"

		}
	},
	{
		name = "array_controller",
		data = {
			type = "Array",
			dataType = "String",
			variableName = "myArray",
			initialValue = "",
			variableFor = "SaveData",
			oneTime = "False",
			arrayLength = "1"
		}
	},
	{
		name = "list_controller",
		data = {
			type = "List",
			dataType = "String",
			variableName = "myList",
			initialValue = "",
			variableFor = "SaveData",
			oneTime = "False",
			arrayLength = "1"
		}
	},
	{
		name = "dictionary_controller",
		data = {
			type = "Dictionary",
			dataType = "String",
			variableName = "myDictionary",
			initialValue = "",
			variableFor = "SaveData",
			oneTime = "False",
			arrayLength = "1"
		}
	}
}

function variableController.ignoredFields(entity)
	local res = consts.hideIDName

	if entity.type == "Array" then
		table.insert(res, "arrayLength")
	end

	return res
end

variableController.fieldOrder = {
	"x", "y",
	"type", "dataType",
	"variableName", "initialValue",
	"variableFor", "oneTime",
	"arrayLength"
}

function variableController.fieldInformation(entity)
	local fields = {
		type = {
			options = vectorHelper.VariableTypes,
			editable = false
		},
		dataType = {
			options = vectorHelper.DataTypes,
			editable = false
		},
		variableFor = {
			options = vectorHelper.SaveMethod,
			editable = false
		},
		oneTime = {
			options = vectorHelper.OneTimeModes,
			editable = false
		}
	}

	return fields
end

function variableController.sprite(room, entity, viewport)
	local sprite = {}
	local x, y = entity.x, entity.y

	local configSprite = drawableSprite.fromTexture("VectorHelper/mapping/config_icon", entity)
	configSprite:setPosition(x - 5, y - 5)
	table.insert(sprite, configSprite)

	local controller = drawableSprite.fromTexture(vectorHelper.ControllerIcon, entity)
	table.insert(sprite, controller)

	-- Default sprite for the variable Type incase something goes wrong
	local typeSprite = drawableSprite.fromTexture("VectorHelper/mapping/unknown", entity)
	typeSprite:setPosition(x + 4, y + 2)

	if entity.type == "Variable" then
		typeSprite = drawableSprite.fromTexture("VectorHelper/mapping/variable_icon", entity)
		typeSprite:setPosition(x + 2, y + 3)
	elseif entity.type == "Array" then
		typeSprite = drawableSprite.fromTexture("VectorHelper/mapping/array_icon", entity)
		typeSprite:setPosition(x, y + 3)
	elseif entity.type == "List" then
		typeSprite = drawableSprite.fromTexture("VectorHelper/mapping/list_icon", entity)
		typeSprite:setPosition(x + 2, y + 4)
	elseif entity.type == "Dictionary" then
		typeSprite = drawableSprite.fromTexture("VectorHelper/mapping/dictionary_icon", entity)
		typeSprite:setPosition(x + 2, y + 4)
	end

	table.insert(sprite, typeSprite)

	return sprite
end

return variableController

File: Ahorn/entities/controllers/VectorHelperVariableController.jl
# File written by SSUnlimited

module VectorHelperVariableController

using ..Ahorn, Maple
using Ahorn.VectorHelper

@mapdef Entity "VectorHelper/VariableController" VariableController(
	x::Integer, y::Integer,
	type::String="Variable", dataType::String="String",
	variableName::String="myVariable", initialValue::String="",
	variableFor::String="SaveData", onlyOnce::String="False",
	setMode::String="OnLevelStart" , arrayLength::String="1"
)

const placements = Ahorn.PlacementDict(
	"Variable Controller (Variables) (Vector Helper)" => Ahorn.EntityPlacement(
		VariableController
	),
	"Array Controller (Variables) (Vector Helper)" => Ahorn.EntityPlacement(
		VariableController,
		"point",
		Dict{String, Any}(
			"type" => "Array",
			"variableName" => "myArray"
		)
	),
	"List Controller (Variables) (Vector Helper)" => Ahorn.EntityPlacement(
		VariableController,
		"point",
		Dict{String, Any}(
			"type" => "List",
			"variableName" => "myList"
		)
	),
	"Dictionary Controller (Variables) (Vector Helper)" => Ahorn.EntityPlacement(
		VariableController,
		"point",
		Dict{String, Any}(
			"type" => "Dictionary",
			"variableName" => "myDictionary"
		)
	)
)

Ahorn.editingOptions(entity::VariableController) = Dict{String, Any}(
	"type" => VectorHelper.VariableTypes,
	"dataType" => VectorHelper.DataTypes,
	"variableFor" => String["SaveData", "Session"],
	"onlyOnce" => VectorHelper.OnlyOnceModes,
	"setMode" => String[ "OnLevelStart", "OnLevelEnd", "OnUpdate" ]
)

Ahorn.editingOrder(entity::VariableController) = String[
	"x", "y",
	"type", "dataType",
	"variableName", "initialValue",
	"variableFor", "onlyOnce",
	"setMode", "arrayLength"
]

function Ahorn.selection(entity::VariableController)
	x, y = Ahorn.position(entity)
	return Ahorn.getSpriteRectangle(VectorHelper.ControllerIcon, x, y)
end

function Ahorn.renderAbs(ctx::Ahorn.Cairo.CairoContext, entity::VariableController, room::Maple.Room)
	x, y = Ahorn.position(entity)

	sprite = VectorHelper.Unknown
	offsetX = 4
	offsetY = 2

    # Change the sprite and offset/position based on the type
	if entity.type == "Variable"
		sprite = "VectorHelper/mapping/variable_icon"
		offsetX = 3
		offsetY = 3
	elseif entity.type == "Array"
		sprite = "VectorHelper/mapping/array_icon"
		offsetX = 1
		offsetY = 3
	elseif entity.type == "List"
		sprite = "VectorHelper/mapping/list_icon"
		offsetX = 3
		offsetY = 4
	elseif entity.type == "Dictionary"
		sprite = "VectorHelper/mapping/dictionary_icon"
		offsetX = 2
		offsetY = 4
	end

	# Draw the gear before the controller sprite so it appears below
	Ahorn.drawSprite(ctx, "VectorHelper/mapping/gear_icon", x - 5, y - 4)
	Ahorn.drawSprite(ctx, VectorHelper.ControllerIcon, x, y)

	# Then the variable type (Variable, Array, etc.)
	Ahorn.drawSprite(ctx, sprite, x + offsetX, y + offsetY)
end

function Ahorn.editingIgnored(entity::VariableController, multiple::Bool=false)
	hide = String[]

    # Hide "arrayLength" if the type is not "Array" so users dont edit it for no reason
	if entity.type != "Array"
		push!(hide, "arrayLength")
	end	

	return hide
end

end